GLTF: Don't duplicate textures when importing blend files

demolke opened the pull request:
blender imports will always start with `res://.godot/imported` because we first convert the blend to gltf, store it in `res://.godot/imported` and run the import from there, so resources linked from blend files end up with duplicate textures after https://github.com/godotengine/godot/pull/96778

Use simplify_path to differentiate between resources actually in the imported folder vs resource relative to GLTF in imported folder.

Introduce a new state where the resource exists on the disk but has not been imported yet.

Fixes: https://github.com/godotengine/godot/issues/98932

fire commented:
I think this makes sense, but what was the original problem being solved?

demolke commented:
This is the issue https://github.com/godotengine/godot/pull/96778#issuecomment-2458415635 

After https://github.com/godotengine/godot/pull/96778 all textures imported from blend files will get duplicated in the res:// folder.

demolke commented:
Ah, me and github - I've uploaded only a part of the change, let me switch it to draft while I try to fix it.

lyuma commented:
Please file a new issue for the bug, with an attached reproduction project, and link to that bug in this PR.

A link to a PR comment makes it difficult to have a meaningful conversation on this issue, and there is enough nuance with this issue that I don't want to merge a PR without an issue

---
I'm missing some information here. For example, the issue purports to affect blend files, but the PR here only touches gltf_document code, which means that the issue must be reproducible from a standalone gltf file. I want to understand what cases are being changed.

Generally speaking, there are two cases for gltf:
1. either the images are embedded, in which case they are extracted with the model name prefixed (`texture` inside `model.gltf` is extracted to `model_texture.png`), or
2. the images are referenced externally from the gltf file, in which case, they are not extracted since they are external references. In this case, `model.gltf` would directly reference `texture.png` which already exists.

This PR is hinting at a possible third case, which I do not understand, in which `model.gltf` (or `model.blend`) references `texture.png` but somehow the gltf_document code also extracts `model_texture.png`.

I am skeptical about this fix, and we should also keep in mind that when blender gltf exports embed a texture file, there is sometimes a reason for it. I believe that, for example, blender will at times make modifications to the texture and embed it to comply with the gltf spec or to handle certain shader nodes.

demolke commented:
The part about blender is somewhat significant, because `.blend` will get converted to `.gltf` and placed in res://.godot/imported - something the https://github.com/godotengine/godot/pull/96778 considers as impossible to import https://github.com/godotengine/godot/blob/77d6283d22731398e23c3f1e1b4eeedd22a93152/modules/gltf/gltf_document.cpp#L4089 so what happens we fail this check because texture linked from a blend file.

(It's very much possible to trigger this if you call gltf import methods yourself with paths relative to res://.godot/imported but no one is probably doing that.)

https://github.com/user-attachments/files/17640095/nonembed.zip

So on disk it looks like this before the image processing

.godot/imported/model.gltf
assets/model.blend
assets/texture.png

The code above will get res://.godot/imported/../../assets/texture.png dismiss it as unimportable, writes the data to the disk res://assets/model-texture.png and imports that.

> This PR is hinting at a possible third case, which I do not understand, in which model.gltf (or model.blend) references texture.png but somehow the gltf_document code also extracts model_texture.png.

Yes, exactly - that's the behavior https://github.com/godotengine/godot/blob/77d6283d22731398e23c3f1e1b4eeedd22a93152/modules/gltf/gltf_document.cpp#L4089 introduced.

I'll port it into an issue.

demolke commented:
https://github.com/godotengine/godot/issues/98932

demolke commented:
The change 
- does `uri.simplify_path();` to make uri absolute
- removes the `p_base_path.begins_with("res://.godot/imported")` check - I think it's meaningless - ResourceLoader will tell you if the texture has been imported or not
- if I've would have left `_parse_image_save_image` as is, all un-imported textures would get read from the disk and then written back for no reason

aaronfranke reviewed (commented):
In the future, please make a new branch for your PRs, instead of using `master`.

Thank you for working on this. It makes sense to separate the `must_write` and `must_import` booleans. And yeah, I completely forgot to consider the case of `res://.godot/imported/../../` which results in things being not inside the `res:://.godot/` folder after all.

However, this doesn't seem to work in practice. PR #96778 was designed to be a GLTF-only PR to make way for PR #96782.

With only PR #96782, the unpacking works correctly (the GLTF material points to the imported texture):

<img width="1270" alt="Screenshot 2024-11-07 at 3 13 28 PM" src="https://github.com/user-attachments/assets/5874e1d6-0096-45e6-ba77-e6a0a92407ae">

With both PR #96782 and this PR, it errors and does not extract (the material's texture is directly loaded an image):

<img width="1270" alt="Screenshot 2024-11-07 at 3 10 35 PM" src="https://github.com/user-attachments/assets/d620dee0-3a15-49c9-af64-d4aa1c635326">

The test file here is the MRP in issue #82455 with "Materials -> Unpack Enabled" set to false (this option should be renamed to "Unpack Originals" in the future).

demolke commented:
@aaronfranke thanks for the pointer. I only tested it with embedded textures and unpack enabled, so will take a look.

demolke commented:
I think I've addressed all the permutations. Added a couple of tests, but they do not capture the full breadth of the problem (that would require blender installation during unit test) - only some of the basic use cases. I am not very happy with them - they write to disk, so are not idempotent. I'm wondering if to include them at all. It would be nice if we have memory backed res:// support.

One interesting thing to note is that there is a race condition in `_parse_image_save_image` with "Materials -> Unpack Enabled" set to true even before https://github.com/godotengine/godot/pull/96778. On first run the blender gltf export will create textures directory which will not be scanned by the time we try to parse it, so `EditorFileSystem::get_singleton()->reimport_append` will not find those files. On the import we will load the textures directly and on the second re-import the folder will be found and the resource linkage succeeds.

demolke commented:
Actually adding the `DirAccess` interface to `FileAccessMemory` wasn't that hard, so have done that to make the tests self-contained. I've left it separated in a different commit for easier review.

demolke commented:
Friendly bump @aaronfranke

aaronfranke commented:
Apologies for the delay. I re-tested the PR and it's the same behavior as I described above here https://github.com/godotengine/godot/pull/98909#pullrequestreview-2422354127

Additionally:
* The changes outside of the GLTF module are out-of-scope and must be made as a separate PR. We can't have a GLTF PR modifying `core/os/os.h`.
* This PR has been modified to include [PR #96782](https://github.com/godotengine/godot/pull/96782), so [PR #96782](https://github.com/godotengine/godot/pull/96782) should be merged first. As far as I know [PR #96782](https://github.com/godotengine/godot/pull/96782) is all good, the bug trying to be fixed here is related but ultimately tangental to it.

demolke commented:
> Apologies for the delay. I re-tested the PR and it's the same behavior as I described above here [#98909 (review)](https://github.com/godotengine/godot/pull/98909#pullrequestreview-2422354127)

Sorry about that - should be fixed now, I've added a test that captures running gltf import from `.godot/imported` to make the handling explicit.

> 
> Additionally:
> 
>     * The changes outside of the GLTF module are out-of-scope and must be made as a separate PR. We can't have a GLTF PR modifying `core/os/os.h`.

Removed the in-memory DirAccess and the test changes.

I need to keep the other changes outside of gltf to

- `editor/*` to handle the memory cleanup to make sanitizers happy. 
- `core/config/engine.h` to add a hint if we're running tests similar to editor hint. The way the texture import happens within existing import for embedded resources is quite hard to set up the same way - see the change to `editor/editor_file_system.cpp`. I wonder if the assert (`!importing`) has to be there in the first place, but it's quite hard to gauge what the side effect would be.

> 
>     * This PR has been modified to include [PR #96782](https://github.com/godotengine/godot/pull/96782), so [PR #96782](https://github.com/godotengine/godot/pull/96782) should be merged first. As far as I know [PR #96782](https://github.com/godotengine/godot/pull/96782) is all good, the bug trying to be fixed here is related but ultimately tangental to it.

Yeah, I was not planning on leaving that in, had it there for testing, removed it now.

> 
> 
> As for the discussion on RocketChat about the importance of fixing this regression, I agree. Worst case scenario we can revert just the second commit in [PR #96778 as described here](https://github.com/godotengine/godot/pull/96778#issuecomment-2446425221), but I think we can find a better solution thanks to your efforts.

aaronfranke commented on modules/gltf/tests/test_gltf_extras.h:
```diff
@@ -212,10 +167,10 @@ TEST_CASE("[SceneTree][Node] GLTF test skeleton and bone export and import") {
 	memdelete(loaded);
 }
-} // namespace TestGltfExtras
+} //namespace TestGltf
```
This seems to be a mistake.

aaronfranke commented on core/config/engine.h:
```diff
@@ -85,6 +85,7 @@ class Engine {
 	HashMap<StringName, Object *> singleton_ptrs;
 
 	bool editor_hint = false;
+	bool tests_hint = false;
```
Can you open a new PR with only the `core/` and `editor/` changes, which can be merged as a prerequisite of this PR?

aaronfranke commented:
I tested this PR and it works correctly now. The code makes sense and a quick review of the test cases looks good. I like how it tests the full editor import pipeline. The call to `unregister_all_gltf_document_extensions` is a bit unfortunate since it removes our ability to unit test extensions like physics from then on, however we can just ensure that any tests for glTF extensions run before the tests added by this PR.

The main thing is that the non-GLTF changes should be in their own PR, including `core/`, `engine/`, and the relevant lines in the `tests/` folder like `Engine::get_singleton()->set_tests_hint(true);`, and then that separate PR should be reviewed by the godotengine/core and godotengine/tests teams.

demolke committed:
GLTF: Don't duplicate textures when importing blend files

Blender imports will always start within `.godot/imported`  folder because we first convert the .blend file to .gltf, store it in `.godot/imported` and run the import from there, so on-disk resources linked from .blend files end up with duplicate textures.

demolke commented:
> The call to `unregister_all_gltf_document_extensions` is a bit unfortunate since it removes our ability to unit test extensions like physics from then on, however we can just ensure that any tests for glTF extensions run before the tests added by this PR.

Actually turns out it's no longer necessary - it was needed back when I introduced the tests, but it's able to find the right nodes after this PR without the unregistration.

> The main thing is that the non-GLTF changes should be in their own PR, including `core/`, `engine/`, and the relevant lines in the `tests/` folder like `Engine::get_singleton()->set_tests_hint(true);`, and then that separate PR should be reviewed by the godotengine/core and godotengine/tests teams.

I did split it now

https://github.com/godotengine/godot/pull/99846
https://github.com/godotengine/godot/pull/99847
https://github.com/godotengine/godot/pull/99848

but it feels the individual PRs lack context and when someone is going to investigate the reason for the changes, they just end up here anyway.

demolke committed:
Move reimport check to EditorImportPlugin

reimport_append is used by gltf_document, fbx_document and editor_import_plugin. The first two will never call it when importing == false. It's only the editor_import_plugin that should guard against that.
https://docs.godotengine.org/en/stable/classes/class_editorimportplugin.html#class-editorimportplugin-method-append-import-external-resource

The motivation of removing the check from gltf_document call path is to be able to test nested imports (texture embedded in gltf).

demolke commented:
Thanks for the review and the suggestions.

akien-mga commented:
Thanks!
